proposals/jib_core_library.md

```diff
@@ -0,0 +1,76 @@
+# Proposal: Jib Core as a library for Java
+
+The tracking issue is at [#337](https://github.com/GoogleContainerTools/jib/issues/337).
+
+# Goal
+
+Design for Jib Core as a Java library for building container images.
+
+# Proposed API
+
+`Jib` - the main entrypoint for using Jib Core
+- `JibContainerBuilder from(String baseImageReference)`
+- `JibContainerBuilder from(ImageReference baseImageReference)`
+- `JibContainerBuilder from(RegistryImage baseImage)`
+
+`JibContainerBuilder` - configures the container to build
+- `JibContainerBuilder layer(List<Path> files, Path pathInContainer)`
+- `JibContainerBuilder layer(LayerConfiguration)`
+- `JibContainerBuilder setLayers(List<LayerConfiguration>/LayerConfiguration...)`
+
+- `JibContainerBuilder setEntrypoint(List<String>/String...)`
+- `JibContainerBuilder setProgramArguments(List<String>/String...)`
+- `JibContainerBuilder setEnvironment(Map<String, String> environmentMap)`
+- `JibContainerBuilder addEnvironmentVariable(String name, String value)`
+- `JibContainerBuilder setExposedPorts(List<Port>/Port...)`
+- `JibContainerBuilder addExposedPort(Port port)`
+- `JibContainerBuilder setLabels(Map<String, String> labelMap)`
+- `JibContainerBuilder addLabel(String key, String value)`
+- `JibContainer containerize(Containerizer)`
+
+Three `TargetImage` types (`RegistryImage`, `DockerDaemonImage`, and `TarImage`) define the 3 different targets Jib can build to.
+
+`RegistryImage` - builds to a container registry
+- `static RegistryImage named(ImageReference/String)`
+- `RegistryImage addCredential(String username, String password)`
+- `RegistryImage addCredentialRetriever(CredentialRetriever)`
+
+`DockerDaemonImage` - builds to a Docker daemon
+- `static DockerDaemonImage named(ImageReference/String)`
+- `DockerDaemonImage setDockerExecutable(Path)`
+
+`TarImage` - builds to a tarball archive
+- `Builder`
+  - `TarImage saveTo(Path outputFile)`
+- `static Builder named(ImageReference/String)`
+
+`Containerizer` - configures how and where to containerize to
```

Comment:
Is `withExectionContext` not a thing anymore?

Reply:
Nope - it made sense to just have the things that would have been under `ExecutionContext` on the `Containerizer` itself since that is the execution context.